docs/guide/api-vite-environment.md

```diff
@@ -0,0 +1,673 @@
+
+A Vite Module Runner allows running any code by processing it with Vite plugins first. It is different from `server.ssrLoadModule` because the runtime implementation is decoupled from the server. This allows library and framework authors to implement their layer of communication between the Vite server and the runner. The browser communicates with its corresponding environment using the server Web Socket and through HTTP requests. The Node Module runner can directly do function calls to process modules as it is running in the same process. Other environments could run modules connecting to an edge runtime like workerd, or a Worker Thread as Vitest does.
+
+All these environments share Vite's HTTP server, middlewares, and Web Socket. The resolved config and plugins pipeline are also shared, but plugins can use `apply` so its hooks are only called for certain environments. The environment can also be accessed inside hooks for fine-grained control.
+
+![Vite Environments](../images/vite-environments.svg)
+
+## Using environments in the Vite server
+
+A Vite dev server exposes two environments by default named `'browser'` and `'node'`. The browser environment runs in client apps that have imported the `/@vite/client` module. The Node environment runs in the same runtime as the Vite server and allows application servers to be used to render requests during dev with full HMR support. We'll discuss later how frameworks and users can create and register new environments.
+
+The available environments can be accessed using the `server.environments` read-only array:
+
+```js
+server.environments.forEach((environment) => log(environment.name))
+```
+
+Each environment can also be accessed through its name:
+
+```js
+server.environment('browser').transformRequest(url)
+```
+
+An environment is an instance of the `ModuleExecutionEnvironment` class:
+
+```ts
+class ModuleExecutionEnvironment {
+  /**
+   * Unique identifier for the environment in a Vite server.
+   * By default Vite exposes 'browser' and 'node' environments.
+   * The ecosystem has consensus on other environments, like 'workerd'.
+   */
+  name: string
+  /**
+   * Communication channel to send and receive messages from the
+   * associated module runner in the target runtime.
+   */
+  hot: HMRChannel | null
+  /**
+   * Graph of module nodes, with the imported relationship between
+   * processed modules and the cached result of the processed code.
+   */
+  moduleGraph: ModuleGraph
+  /**
+   * Trigger the execution of a module using the associated module runner
+   * in the target runtime.
+   */
+  run: ModuleRunFunction
+  /**
+   * Resolved config options for this environment. Options at the server
+   * global scope are taken as defaults for all environments, and can
+   * be overridden (resolve conditions, external, optimizedDeps)
+   */
+  config: ResolvedEnvironmentConfig
+
+  constructor(
+    server,
+    { name, hot, run, config }: ModuleExecutionEnvironmentOptions,
```

Comment:
`run` here isn't the module runner but a function that connects the ModuleExecutionEnvironment instance to the ModuleRunner (possibly in another runtime as you said). This function will use function calls or RPC to be able to execute code in the runner and get the result.

The `run` function at first was a way to trigger execution, but given that we want to abstract the `ssrEnvironment`, we need to also be able to get the results and interact with them (your [comment here](https://github.com/vitejs/vite/pull/16089#discussion_r1515462163)). @sheremet-va had some ideas that we still need to explore and write down, that if I understand correctly would allow us to avoid the need for environments to specify `run` but rather implement some more low level transport layer so we can implement it in Vite core directly.